### PR TITLE
Curve testing

### PR DESCRIPTION
**Pull request by franalbani**
This PR adds Curve testing for all functions, excluding those that had errors initially.

It also makes changes in `Curve.py`:

* `float` --> `Decimal` #50 
* fix shape discrepancy #57 
* new retry logic #18 
* no more bare `Exception`

### CODE REVIEW

File: defi_protocols/Curve.py [hunk 1]
```diff
@@ -150,122 +134,102 @@ def get_pool_gauge_address(web3, pool_address, lptoken_address, block, blockchai
...
-        except:
-            pass
+                return ['ChildGauge', gauge_contract]
+
+        if only_version is True:
+            return 'LiquidityGaugeV5'
+        else:
+            return ['LiquidityGaugeV5', gauge_contract]
+    except Exception as exception:
+        logger.error(f'{exception = }')
```
Comment (sanekto) [inline defi_protocols/Curve.py:164]: why not logger.exception?

Comment (franalbani) [reply defi_protocols/Curve.py:164]: Because its output is huge.

File: defi_protocols/Curve.py [hunk 2]
```diff
@@ -150,122 +134,102 @@ def get_pool_gauge_address(web3, pool_address, lptoken_address, block, blockchai
...
             if only_version is True:
-                return 'LiquidityGaugeV5'
+                return 'ChildGauge'
             else:
-                return ['LiquidityGaugeV5', gauge_contract]
-        except:
-            pass
+                return ['ChildGauge', gauge_contract]
+
+        if only_version is True:
```
Comment (sanekto) [inline defi_protocols/Curve.py:159]: The pythonic way is to `if only_version:`

File: defi_protocols/Curve.py (hunk 1 above)
Comment (sanekto) [reply defi_protocols/Curve.py:164]: Ahhhhh so the logging is some debugging? not something to report to the end user right?

File: tests/test_curve.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+import logging
+
+from defi_protocols.functions import get_node
+from defi_protocols.constants import ETHEREUM, X3CRV_ETH, CRV_ETH, DAI_ETH, USDC_ETH, USDT_ETH
+from defi_protocols import Curve, add_stderr_logger
+
+
+_h = add_stderr_logger(logging.DEBUG)
+logger = logging.getLogger(__name__)
+logger.addHandler(_h)
```
Comment (sanekto) [inline tests/test_curve.py:11]: The logger configuration inside each test file should not be needed. Use `pytest --log-cli-level=debug` for example.

File: defi_protocols/Curve.py (hunk 1 above)
Comment (franalbani) [reply defi_protocols/Curve.py:164]: Exactly. I used them to help me understand what code was being exercised.

Until a better refactor, I will catch only the exception that "flows-control" and remove that line.

Comment (sanekto) [reply defi_protocols/Curve.py:164]: I would use logger.debug for that, so when I run the code and I see an error it is an actual error (not this)

File: tests/test_curve.py (hunk 3 above)
Comment (franalbani) [reply tests/test_curve.py:11]: Oh.. I was using only `--log-level` and missing them.

Comment (franalbani) [reply tests/test_curve.py:11]: I tried `--log-cli-level` and now I'm inundated with messages from a lot of libs.

Comment (sanekto) [reply tests/test_curve.py:11]: I sometimes use a `conftest.py` file, placed in the root of the repo, to be able to specify more detailed things.
What about having something like this to disable all the other loggers that are not 'defi_protocols'?
```
import logging

def pytest_addoption(parser):
    parser.addoption(
        "--defi-debug-only",  action="store_true", default=False, help="enable defi-protocols debug logging"
    )

def pytest_configure(config):
    if config.getoption("--defi-debug-only"):
        for name, logger in logging.root.manager.loggerDict.items():
            if "defi_protocols" not in name:
                logger.level = logging.WARNING
```

And then execute with `pytest  -vs --log-cli-level=debug --defi-debug-only  ...`

**Comment by franalbani** [comment]
Disregarding the functions I could not test because of errors, I consider this branch ready.

**Pull request merged**